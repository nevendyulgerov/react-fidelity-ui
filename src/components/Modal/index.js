import Modal from './script/Modal';

export default Modal