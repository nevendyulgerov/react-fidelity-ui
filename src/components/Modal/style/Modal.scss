/**
 * Component: Modal
 */

[data-component="modal"] {
  opacity: 0;
  pointer-events: none;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 999;
  display: flex;
  justify-content: center;
  align-items: flex-start;
  transition: all 0.15s ease-in-out;

  &.active {
    opacity: 1;
    pointer-events: all;
    transition: all 0.15s ease-in-out;
    background-color: $transparent-grey;
  }

  > .component-body {
    position: relative;
    top: 40px;
    width: 60%;
    height: auto;
    background-color: $white;
    box-shadow: $shadow-depth-2;
    border-radius: 4px;
    overflow: hidden;

    @media (max-width: $screen-md) {
      width: 75%;
    }

    @media (max-width: $screen-sm) {
      width: 85%;
    }

    @media (max-width: $screen-xs) {
      width: 90%;
    }

    @media (max-width: $screen-xxs) {
      width: 90%;
    }
  }
}
