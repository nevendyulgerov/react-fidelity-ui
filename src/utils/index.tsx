/**
 * @description Is null
 * @param val
 */
export const isNull = (val: any) : boolean => val === null;

/**
 * @description Is array
 * @param val
 */
export const isArr = (val: any) : boolean => Array.isArray(val);

/**
 * @description Is object
 * @param val
 */
export const isObj = (val: any) : boolean => typeof val === 'object' && !isArr(val) && !isNull(val);

/**
 * @description Is number
 * @param val
 */
// eslint-disable-next-line no-restricted-globals
export const isNum = (val: any) : boolean => typeof val === 'number' && !isNaN(val);

/**
 * @description Is function
 * @param val
 */
export const isFunc = (val: any) : boolean => typeof val === 'function';

/**
 * @description Is string
 * @param val
 */
export const isStr = (val: any) : boolean => typeof val === 'string';

/**
 * @description Is undefined
 * @param val
 */
export const isUndef = (val: any) : boolean => typeof val === 'undefined';

/**
 * @description Is boolean
 * @param val
 */
export const isBool = (val: any) : boolean => typeof val === 'boolean';

/**
 * @description Is non empty string
 * @param val
 */
export const isNonEmptyStr = (val: any) : boolean => isStr(val) && val !== '';

/**
 * @description Uid
 * @param len
 */
export const uid = (len: number = 7): string => Math.random().toString(35).substr(2, len);

/**
 * @description Get object keys
 * @param obj
 * @returns {Array}
 */
export const getKeys = (obj: any) => Object.keys(obj);

/**
 * @description Each key
 * @param obj
 * @param callback
 */
export const eachKey = (obj: any, callback: (key: string, prop: any, index: number) => void) => {
  getKeys(obj).forEach((key, index) => callback(key, obj[key], index));
};

/**
 * @description Observe intersection
 * @param target
 * @param callback
 */
export const observeIntersection = (target: HTMLElement, callback: (items: []) => void) => {
  if (!isFunc(window.IntersectionObserver)) {
    return console.warn('IntersectionObserver is not supported.');
  }

  // eslint-disable-next-line compat/compat
  const Observe = new IntersectionObserver((items: []) => {
    callback(items);
  });

  Observe.observe(target);

  return Observe;
};

/**
 * @description Get css var
 * @param varName
 */
export const getCssVar = (varName: string): string | null => (
  getComputedStyle(document.documentElement).getPropertyValue(varName)
);

/**
 * @description Set css var
 * @param varName
 * @param varValue
 */
export const setCssVar = (varName: string, varValue: string | null) => {
  document.documentElement.style.setProperty(varName, varValue);
};

/**
 * @description Limit number
 * @param num
 * @param max
 */
export const limitNum = (num: number, max: number) => (num <= max ? num : max);

/**
 * @description Can render prop
 * @param prop
 */
export const isRenderable = (prop: any) => isNonEmptyStr(prop) || isNum(prop) || isObj(prop);
