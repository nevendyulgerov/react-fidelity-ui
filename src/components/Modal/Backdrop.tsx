import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import { isNonEmptyStr } from '../../utils';

export type ModalBackdropProps = {
  children: React.ReactNode,
  className?: string | null,
  [key: string]: any
};

const ModalBackdrop = (props: ModalBackdropProps) => {
  const { children, className, ...restProps } = props;
  const componentClassName = classNames({
    modal__backdrop: true,
    // @ts-ignore
    [className]: isNonEmptyStr(className)
  });

  return (
    <div
      role="document"
      {...restProps}
      className={componentClassName}
    >
      {children}
    </div>
  );
};

ModalBackdrop.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]).isRequired,
  className: PropTypes.string
};

ModalBackdrop.defaultProps = {
  className: null
};

export default ModalBackdrop;
