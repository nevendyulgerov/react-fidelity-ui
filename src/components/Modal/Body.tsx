import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import { isNonEmptyStr } from '../../utils';

export type ModalBodyProps = {
  children: React.ReactNode,
  className?: string | null,
  [key: string]: any
};

const ModalBody = (props: ModalBodyProps) => {
  const { children, className, ...restProps } = props;
  const componentClassName = classNames({
    modal__body: true,
    // @ts-ignore
    [className]: isNonEmptyStr(className)
  });

  return (
    <div
      {...restProps}
      className={componentClassName}
    >
      {children}
    </div>
  );
};

ModalBody.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]).isRequired,
  className: PropTypes.string
};

ModalBody.defaultProps = {
  className: null
};

export default ModalBody;
