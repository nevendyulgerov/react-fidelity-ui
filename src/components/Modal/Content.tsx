import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import { isNonEmptyStr } from '../../utils';

export type ModalContentProps = {
  children: React.ReactNode,
  size?: string | 'auto',
  className?: string | null,
  [key: string]: any
};

const sizes = ['auto', 'xs', 'sm', 'md', 'lg', 'xl', 'fullscreen'];

const ModalContent = (props: ModalContentProps) => {
  const { children, size, className, ...restProps } = props;
  const componentClassName = classNames({
    modal__content: true,
    // @ts-ignore
    [`modal__content--size-${size}`]: ModalContent.sizes.includes(size),
    // @ts-ignore
    [className]: isNonEmptyStr(className)
  });

  return (
    <div
      {...restProps}
      className={componentClassName}
    >
      {children}
    </div>
  );
};

ModalContent.sizes = sizes;

ModalContent.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]).isRequired,
  size: PropTypes.oneOf(sizes),
  className: PropTypes.string
};

ModalContent.defaultProps = {
  size: 'auto',
  className: null
};

export default ModalContent;
