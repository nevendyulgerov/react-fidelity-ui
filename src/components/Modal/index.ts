import Body from './Body';
import Footer from './Footer';
import Header from './Header';
import Modal from './Modal';
import Title from './Title';

export {
  Body,
  Footer,
  Header,
  Modal,
  Title
};
