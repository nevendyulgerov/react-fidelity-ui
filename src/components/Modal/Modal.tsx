import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import { disableBodyScroll, clearAllBodyScrollLocks } from 'body-scroll-lock';
import Backdrop from './Backdrop';
import Body from './Body';
import Content from './Content';
import Footer from './Footer';
import Header from './Header';
import { isNonEmptyStr } from '../../utils';

const { useRef, useEffect } = React;

export type ModalProps = {
  children: React.ReactNode,
  className?: string | null,
  active?: boolean | false,
  disableScroll?: boolean | true,
  [key: string]: any
};

const Modal = (props: ModalProps) => {
  const { children, className, active, disableScroll, ...restProps } = props;
  const refComponent = useRef(null);
  const componentClassName: string = classNames({
    modal: true,
    'modal--active': active,
    // @ts-ignore
    [className]: isNonEmptyStr(className)
  });

  useEffect(() => {
    if (disableScroll && active) {
      const modalScroll = refComponent.current;
      // @ts-ignore
      disableBodyScroll(modalScroll);
    }

    if (disableScroll && !active) {
      clearAllBodyScrollLocks();
    }

    return () => {
      if (disableScroll) {
        clearAllBodyScrollLocks();
      }
    };
  }, [disableScroll, active]);

  return (
    <div
      role="dialog"
      tabIndex={-1}
      aria-hidden={!active}
      {...restProps}
      ref={refComponent}
      className={componentClassName}
    >
      {children}
    </div>
  );
};

Modal.Backdrop = Backdrop;
Modal.Body = Body;
Modal.Content = Content;
Modal.Footer = Footer;
Modal.Header = Header;

Modal.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]).isRequired,
  className: PropTypes.string,
  active: PropTypes.bool,
  disableScroll: PropTypes.bool
};

Modal.defaultProps = {
  className: null,
  active: false,
  disableScroll: true
};

export default Modal;
