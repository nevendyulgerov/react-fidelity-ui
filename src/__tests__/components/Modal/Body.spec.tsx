import * as React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Modal from '../../../components/Modal';
import { isObj } from '../../../utils';

describe('Components / Modal / Body', () => {
  it('matches snapshot', () => {
    const { baseElement } = render(
      <Modal.Body>
        Modal Body
      </Modal.Body>
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('displays children', () => {
    const { baseElement } = render(
      <Modal.Body>
        <h1>Heading 1</h1>
        <h2>Heading 2</h2>
      </Modal.Body>
    );
    const node = baseElement.querySelector('.modal__body');

    // @ts-ignore
    expect(node.children.length).toBe(2);
  });

  it('attaches given className', () => {
    const className = 'modal__body-xyz';
    const { baseElement } = render(
      <Modal.Body className={className}>
        Modal Body
      </Modal.Body>
    );
    const node = baseElement.querySelector(`.${className}`);

    expect(isObj(node)).toBe(true);
  });

  it('adds given style prop', () => {
    const style = { height: '20px' };
    const { baseElement } = render(
      <Modal.Body style={style}>
        Modal Body
      </Modal.Body>
    );
    const node = baseElement.querySelector('.modal__body');

    // @ts-ignore
    expect(getComputedStyle(node).height).toBe(style.height);
  });

  it('adds given callback prop', () => {
    let isClickTriggered = false;
    const onClick = () => {
      isClickTriggered = true;
    };
    const { baseElement } = render(
      <Modal.Body onClick={onClick}>
        Modal Body
      </Modal.Body>
    );
    const node = baseElement.querySelector('.modal__body');

    // @ts-ignore
    fireEvent.click(node);

    expect(isClickTriggered).toBe(true);
  });

  it('adds given role prop', () => {
    const role = 'modal-body';
    const { baseElement } = render(
      <Modal.Body role={role}>
        Modal Body
      </Modal.Body>
    );
    const node = baseElement.querySelector('.modal__body');

    // @ts-ignore
    expect(node.getAttribute('role')).toBe(role);
  });
});
