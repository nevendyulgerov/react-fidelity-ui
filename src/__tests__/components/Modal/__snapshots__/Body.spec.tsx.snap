// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / Modal / Body matches snapshot 1`] = `
<body>
  <div>
    <div
      class="modal__body"
    >
      Modal Body
    </div>
  </div>
</body>
`;
